File: examples/counter/cli/Sources/CounterE2E.swift
import Foundation
import JavaCardRPCClient
import CounterClient

// MARK: - DI: Conform TCPTransport to generated CounterTransport protocol

extension TCPTransport: @retroactive CounterTransport {
    public func transmit(cla: UInt8, ins: UInt8, p1: UInt8, p2: UInt8, data: Data?) async throws -> (sw: UInt16, data: Data) {
        let command = APDUCommand(cla: cla, ins: ins, p1: p1, p2: p2, data: data)
        let response = try await transmit(command)
        return (sw: response.sw, data: response.data)
    }
}

// MARK: - E2E Test Runner

@main
struct CounterE2E {
    static func main() async {
        let transport = TCPTransport(host: "127.0.0.1", port: 9025)

        print("=== javacard-rpc Counter E2E Tests ===\n")

        do {
            try transport.connect()
            print("[OK] Connected to bridge\n")
        } catch {
            print("[FAIL] Connection failed: \(error)")
            exit(1)
        }

        let counter = CounterClient(transport: transport)
        var passed = 0
        var failed = 0

        func test(_ name: String, _ body: () async throws -> Void) async {
            do {
                try await body()
                print("[PASS] \(name)")
                passed += 1
            } catch {
                print("[FAIL] \(name): \(error)")
                failed += 1
            }
        }

        func expect<T: Equatable>(_ actual: T, _ expected: T, _ msg: String) throws {
            guard actual == expected else {
                throw TestError.assertion("\(msg): expected \(expected), got \(actual)")
            }
        }

        func expectData(_ actual: Data, _ expected: Data, _ msg: String) throws {
            guard actual == expected else {
                throw TestError.assertion("\(msg): expected \(expected as NSData), got \(actual as NSData)")
            }
        }

        func parseTLV(_ data: Data) throws -> [UInt8: Data] {
            var out: [UInt8: Data] = [:]
            var offset = 0
            while offset < data.count {
                guard offset + 2 <= data.count else {
                    throw TestError.assertion("truncated TLV header")
                }
                let tag = data[offset]
                let len = Int(data[offset + 1])
                offset += 2
                guard offset + len <= data.count else {
                    throw TestError.assertion("truncated TLV value")
                }
                out[tag] = data.subdata(in: offset..<(offset + len))
                offset += len
            }
            return out
        }

        func expectPrefix(_ actual: Data, _ prefix: [UInt8], _ msg: String) throws {
            let expected = Data(prefix)
            guard actual.count >= expected.count, actual.prefix(expected.count) == expected else {
                throw TestError.assertion("\(msg): wrong prefix")
            }
        }

        func expectedMockSignature(challenge: Data, counter: UInt16, limit: UInt16) -> Data {
            let partLen = min(challenge.count, 8)
            var out = Data(count: 2 + 2 + partLen + 2 + partLen)
            out[0] = 0x30
            out[1] = UInt8(out.count - 2)
            out[2] = 0x02
            out[3] = UInt8(partLen)
            for i in 0..<partLen {
                out[4 + i] = challenge[i] ^ 0x01 ^ UInt8(counter & 0x00FF)
            }
            out[4] &= 0x7F

            let secondOff = 4 + partLen
            out[secondOff] = 0x02
            out[secondOff + 1] = UInt8(partLen)
            for i in 0..<partLen {
                let srcIndex = challenge.count - 1 - i
                out[secondOff + 2 + i] = challenge[srcIndex] ^ UInt8(limit & 0x00FF) ^ 0x5A
            }
            out[secondOff + 2] &= 0x7F
            return out
        }

        func expectSW(_ body: () async throws -> Void, sw: UInt16) async throws {
            do {
                try await body()
                throw TestError.assertion("Expected SW \(String(format: "%04X", sw)) but got success")
            } catch is TestError {
                throw TestError.assertion("Expected SW \(String(format: "%04X", sw)) but got success")
            } catch {
                // Generated client throws TransportError.statusWord(UInt16)
                // Check the error description contains the expected SW value
                let desc = String(describing: error)
                let swDecimal = String(sw)
                guard desc.contains(swDecimal) || desc.contains(String(format: "%04X", sw).lowercased()) else {
                    throw TestError.assertion("Expected SW \(String(format: "%04X", sw)), got error: \(error)")
                }
            }
        }

        // Test 1: SELECT applet
        await test("1. SELECT applet") {
            try await counter.select()
        }

        // Test 2: increment(5) → 5, increment(3) → 8
        await test("2. increment(5) → 5, increment(3) → 8") {
            let v1 = try await counter.increment(amount: 5)
            try expect(v1, UInt16(5), "increment(5)")
            let v2 = try await counter.increment(amount: 3)
            try expect(v2, UInt16(8), "increment(3)")
        }

        // Test 3: get() → 8
        await test("3. get() → 8") {
            let v = try await counter.get()
            try expect(v, UInt16(8), "get()")
        }

        // Test 4: decrement(2) → 6
        await test("4. decrement(2) → 6") {
            let v = try await counter.decrement(amount: 2)
            try expect(v, UInt16(6), "decrement(2)")
        }

        // Test 5: decrement(100) → SW_UNDERFLOW
        await test("5. decrement(100) → SW_UNDERFLOW") {
            try await expectSW({ try await counter.decrement(amount: 100) }, sw: 0x6985)
        }

        // Test 6: setLimit(10), increment(5) → SW_OVERFLOW (6+5=11 > 10)
        await test("6. setLimit(10), increment(5) → SW_OVERFLOW") {
            try await counter.setLimit(limit: 10)
            try await expectSW({ try await counter.increment(amount: 5) }, sw: 0x6986)
        }

        // Test 7: store("hello world"), load() → "hello world"
        await test("7. store + load roundtrip") {
            let data = "hello world".data(using: .utf8)!
            try await counter.store(data: data)
            let loaded = try await counter.load()
            guard loaded == data else {
                throw TestError.assertion("store/load mismatch: expected \(data.count) bytes, got \(loaded.count)")
            }
        }

        // Test 8: load() before store → SW_NO_DATA (need new session)
        // Skip — we already stored data. Tested in jCardSim tests.
        // Instead, test store with too-large data
        await test("8. store(200 bytes) → SW_DATA_TOO_LONG") {
            let bigData = Data(repeating: 0xAB, count: 200)
            try await expectSW({ try await counter.store(data: bigData) }, sw: 0x6A80)
        }

        // Test 9: getInfo() → correct values and flags
        await test("9. getInfo() → {value=6, limit=10, version=1, hasStoredData=true, isAtLimit=false}") {
            let info = try await counter.getInfo()
            try expect(info.value, UInt16(6), "value")
            try expect(info.limit, UInt16(10), "limit")
            try expect(info.version, UInt8(1), "version")
            try expect(info.hasStoredData, true, "hasStoredData")
            try expect(info.isAtLimit, false, "isAtLimit")
        }

        // Test 10: getSpki() → 91-byte mock DER
        await test("10. getSpki() → fixed 91-byte mock DER") {
            let spki = try await counter.getSpki()
            try expect(spki.count, 91, "spki byte count")
            try expectPrefix(spki, [0x30, 0x59, 0x30, 0x13], "spki")
            try expect(spki[26], 0x04, "spki EC point prefix")
        }

        // Test 11: getImsi() → ASCII digits
        await test("11. getImsi() → ASCII digits") {
            let imsi = try await counter.getImsi()
            try expectData(imsi, Data("250011234567890".utf8), "imsi")
        }

        // Test 12: getAppletInfo() → mock TLV payload
        await test("12. getAppletInfo() → expected TLV payload") {
            let info = try await counter.getAppletInfo()
            let tlv = try parseTLV(info)
            try expectData(tlv[0x01] ?? Data(), Data([0x01]), "schemaVersion")
            try expectData(tlv[0x02] ?? Data(), Data([0xF0, 0x00, 0x00, 0x01, 0x01]), "appletAid")
            try expectData(tlv[0x03] ?? Data(), Data("1.0.0".utf8), "appletVersion")
            try expectData(tlv[0x04] ?? Data(), Data([0x01]), "keyAlgorithm")
            try expectData(tlv[0x05] ?? Data(), Data([0x00, 0x3F]), "capabilities")
        }

        // Test 13: signChallenge() → deterministic mock DER
        await test("13. signChallenge() → deterministic mock DER") {
            let challenge = Data("auth-like-challenge".utf8)
            let signature = try await counter.signChallenge(challenge: challenge)
            let info = try await counter.getInfo()
            let expected = expectedMockSignature(challenge: challenge, counter: info.value, limit: info.limit)
            try expectData(signature, expected, "signature")
        }

        // Test 14: signChallenge(empty) → SW_EMPTY_CHALLENGE
        await test("14. signChallenge(empty) → SW_EMPTY_CHALLENGE") {
            try await expectSW({ try await counter.signChallenge(challenge: Data()) }, sw: 0x6700)
        }

        // Test 15: reset() → get() → 0
        await test("15. reset() → get() → 0") {
            try await counter.reset()
            let v = try await counter.get()
            try expect(v, UInt16(0), "get after reset")
        }

        // Test 16: store(128 bytes) → OK, load() → same
        await test("16. store(128 bytes max) → load() roundtrip") {
            let data = Data((0..<128).map { UInt8($0) })
            try await counter.store(data: data)
            let loaded = try await counter.load()
            guard loaded == data else {
                throw TestError.assertion("128-byte roundtrip failed")
            }
        }

        print("\n=== Results: \(passed) passed, \(failed) failed ===")
        transport.disconnect()
        exit(failed > 0 ? 1 : 0)
    }
}

enum TestError: Error, CustomStringConvertible {
    case assertion(String)
    var description: String {
        switch self { case .assertion(let msg): return msg }
    }
}

File: examples/counter/kotlin-cli/src/main/kotlin/io/jcrpc/example/CounterE2E.kt
package io.jcrpc.example

import counter.CounterAppletInfo
import counter.CounterClient
import counter.CounterClientException
import counter.CounterInfo
import counter.CounterTransport
import counter.CounterTransportResult
import io.jcrpc.client.APDUCommand
import io.jcrpc.client.APDUTransport
import io.jcrpc.client.TCPTransport
import kotlinx.coroutines.runBlocking
import kotlin.system.exitProcess

private const val DISPLAY_NAME = "\u041F\u0440\u0438\u0432\u0435\u0442, BSim"
private const val ROUNDTRIP_MESSAGE = "\u041F\u0440\u0438\u0432\u0435\u0442, BSim \uD83D\uDE80"

private class CounterBridgeTransport(
    private val transport: APDUTransport,
) : CounterTransport {
    override suspend fun transmit(
        cla: UByte,
        ins: UByte,
        p1: UByte,
        p2: UByte,
        data: ByteArray?,
    ): CounterTransportResult {
        val response = transport.transmit(APDUCommand(cla = cla, ins = ins, p1 = p1, p2 = p2, data = data))
        return CounterTransportResult(sw = response.sw, data = response.data)
    }
}

fun main(): Unit = runBlocking {
    val tcpTransport = TCPTransport(host = "127.0.0.1", port = 9025)

    println("=== javacard-rpc Counter Kotlin E2E Tests ===")
    println()

    try {
        tcpTransport.connect()
        println("[OK] Connected to bridge")
        println()
    } catch (error: Exception) {
        println("[FAIL] Connection failed: $error")
        exitProcess(1)
    }

    val counter = CounterClient(transport = CounterBridgeTransport(tcpTransport))
    var passed = 0
    var failed = 0

    suspend fun test(name: String, body: suspend () -> Unit) {
        try {
            body()
            println("[PASS] $name")
            passed += 1
        } catch (error: Exception) {
            println("[FAIL] $name: $error")
            failed += 1
        }
    }

    fun <T> expect(actual: T, expected: T, message: String) {
        if (actual != expected) {
            throw IllegalStateException("$message: expected $expected, got $actual")
        }
    }

    fun expectBytes(actual: ByteArray, expected: ByteArray, message: String) {
        if (!actual.contentEquals(expected)) {
            throw IllegalStateException("$message: expected ${expected.toHexString()}, got ${actual.toHexString()}")
        }
    }

    fun expectPrefix(actual: ByteArray, prefix: ByteArray, message: String) {
        if (actual.size < prefix.size || !actual.copyOfRange(0, prefix.size).contentEquals(prefix)) {
            throw IllegalStateException("$message: wrong prefix")
        }
    }

    fun expectedMockSignature(challenge: ByteArray, counter: UShort, limit: UShort): ByteArray {
        val partLen = minOf(challenge.size, 8)
        val output = ByteArray(2 + 2 + partLen + 2 + partLen)
        output[0] = 0x30
        output[1] = (output.size - 2).toByte()
        output[2] = 0x02
        output[3] = partLen.toByte()

        for (index in 0 until partLen) {
            output[4 + index] = (challenge[index].toInt() xor 0x01 xor (counter.toInt() and 0xFF)).toByte()
        }
        output[4] = (output[4].toInt() and 0x7F).toByte()

        val secondOffset = 4 + partLen
        output[secondOffset] = 0x02
        output[secondOffset + 1] = partLen.toByte()
        for (index in 0 until partLen) {
            val sourceIndex = challenge.lastIndex - index
            output[secondOffset + 2 + index] =
                (challenge[sourceIndex].toInt() xor (limit.toInt() and 0xFF) xor 0x5A).toByte()
        }
        output[secondOffset + 2] = (output[secondOffset + 2].toInt() and 0x7F).toByte()
        return output
    }

    suspend fun expectSW(expectedSW: UShort, body: suspend () -> Unit) {
        try {
            body()
            throw IllegalStateException("Expected SW ${expectedSW.toHexString()} but got success")
        } catch (error: CounterClientException.StatusWord) {
            if (error.sw != expectedSW) {
                throw IllegalStateException("Expected SW ${expectedSW.toHexString()}, got ${error.sw.toHexString()}")
            }
        }
    }

    test("1. SELECT applet") {
        counter.select()
    }

    test("2. increment(5) -> 5, increment(3) -> 8") {
        expect(counter.increment(5u), 5u.toUShort(), "increment(5)")
        expect(counter.increment(3u), 8u.toUShort(), "increment(3)")
    }

    test("3. get() -> 8") {
        expect(counter.get(), 8u.toUShort(), "get()")
    }

    test("4. decrement(2) -> 6") {
        expect(counter.decrement(2u), 6u.toUShort(), "decrement(2)")
    }

    test("5. decrement(100) -> SW_UNDERFLOW") {
        expectSW(0x6985u.toUShort()) { counter.decrement(100u) }
    }

    test("6. setLimit(10), increment(5) -> SW_OVERFLOW") {
        counter.setLimit(10u)
        expectSW(0x6986u.toUShort()) { counter.increment(5u) }
    }

    test("7. store + load roundtrip") {
        val data = "hello world".encodeToByteArray()
        counter.store(data)
        expectBytes(counter.load(), data, "store/load roundtrip")
    }

    test("8. store(200 bytes) -> SW_DATA_TOO_LONG") {
        val largePayload = ByteArray(200) { 0xAB.toByte() }
        expectSW(0x6A80u.toUShort()) { counter.store(largePayload) }
    }

    test("9. getInfo() -> typed state snapshot") {
        val info: CounterInfo = counter.getInfo()
        expect(info.value, 6u.toUShort(), "value")
        expect(info.limit, 10u.toUShort(), "limit")
        expect(info.version, 1u.toUByte(), "version")
        expect(info.hasStoredData, true, "hasStoredData")
        expect(info.isAtLimit, false, "isAtLimit")
    }

    test("10. getSpki() -> fixed 91-byte mock DER") {
        val spki = counter.getSpki()
        expect(spki.size, 91, "spki byte count")
        expectPrefix(spki, byteArrayOf(0x30, 0x59, 0x30, 0x13), "spki")
        expect(spki[26].toUByte(), 0x04u.toUByte(), "spki EC point prefix")
    }

    test("11. getImsi() -> ASCII digits") {
        expect(counter.getImsi(), "250011234567890", "imsi")
    }

    test("12. getAppletInfo() -> typed metadata payload") {
        val info: CounterAppletInfo = counter.getAppletInfo()
        expect(info.schemaVersion, 0x01u.toUByte(), "schemaVersion")
        expectBytes(info.appletAid, byteArrayOf(0xF0.toByte(), 0x00, 0x00, 0x01, 0x01), "appletAid")
        expect(info.versionMajor, 0x01u.toUByte(), "versionMajor")
        expect(info.versionMinor, 0x00u.toUByte(), "versionMinor")
        expect(info.versionPatch, 0x00u.toUByte(), "versionPatch")
        expect(info.keyAlgorithm, 0x01u.toUByte(), "keyAlgorithm")
        expect(info.capabilities, 0x003Fu.toUShort(), "capabilities")
    }

    test("13. signChallenge() -> deterministic mock DER") {
        val challenge = "auth-like-challenge".encodeToByteArray()
        val signature = counter.signChallenge(challenge)
        val info = counter.getInfo()
        val expectedSignature = expectedMockSignature(challenge = challenge, counter = info.value, limit = info.limit)
        expectBytes(signature, expectedSignature, "signature")
    }

    test("14. signChallenge(empty) -> SW_EMPTY_CHALLENGE") {
        expectSW(0x6700u.toUShort()) { counter.signChallenge(byteArrayOf()) }
    }

    test("15. getDisplayName() -> UTF-8 string") {
        expect(counter.getDisplayName(), DISPLAY_NAME, "displayName")
    }

    test("16. echoMessage() -> UTF-8 roundtrip") {
        expect(counter.echoMessage(ROUNDTRIP_MESSAGE), ROUNDTRIP_MESSAGE, "echoMessage")
    }

    test("17. reset() -> get() -> 0") {
        counter.reset()
        expect(counter.get(), 0u.toUShort(), "get after reset")
    }

    test("18. store(128 bytes max) -> load() roundtrip") {
        val data = ByteArray(128) { index -> index.toByte() }
        counter.store(data)
        expectBytes(counter.load(), data, "128-byte roundtrip")
    }

    println()
    println("=== Results: $passed passed, $failed failed ===")
    tcpTransport.disconnect()
    exitProcess(if (failed > 0) 1 else 0)
}

private fun UShort.toHexString(): String = "0x%04X".format(this.toInt())

private fun ByteArray.toHexString(): String = joinToString(separator = "") { "%02X".format(it.toUByte().toInt()) }
